pytorch_lightning/plugins/sharded_plugin.py

```diff
@@ -0,0 +1,95 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import platform
+from typing import List, Optional, Union
+
+from pytorch_lightning.core.lightning import LightningModule
+from pytorch_lightning.plugins.ddp_plugin import DDPPlugin
+from pytorch_lightning.utilities import rank_zero_only, _module_available
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+
+if platform.system() != "Windows" and _module_available('fairscale.optim'):  # Distributed not supported on windows
+    from fairscale.optim import OSS
+    from pytorch_lightning.overrides.fairscale import LightningShardedDataParallel
+
+    FAIRSCALE_AVAILABLE = True
+else:
+    FAIRSCALE_AVAILABLE = False
+
+
+class DDPShardedPlugin(DDPPlugin):
+
+    def __init__(self, **kwargs):
+        self._check_fairscale()
+        super().__init__(**kwargs)
+
+    def configure_ddp(
+            self, model: LightningModule, device_ids: List[int]
+    ):
+        self._wrap_optimizers(model)
+        if model.trainer.testing:  # Revert to standard DDP if testing
+            return super().configure_ddp(
+                model=model,
+                device_ids=device_ids
+            )
+        return LightningShardedDataParallel(model, sharded_optimizer=model.trainer.optimizers)
+
+    def optimizer_state(self, optimizer: 'OSS') -> Optional[dict]:
+        optimizer.consolidate_state_dict()
+        return self._optim_state_dict(optimizer)
+
+    def on_before_forward(self, model: LightningModule, *args):
+        args = list(args)
+        batch = args[0]
```

Comment:
this is tricky, if you assume that the first is always batch, lets define it in the method and the later are args